README.md
PHP-VAST
========

forked from sokil/vast (https://github.com/sokil/php-vast.git)

VAST Ad generator and parser library on PHP.

## Specs
* VAST 2.0 Spec: http://www.iab.net/media/file/VAST-2_0-FINAL.pdf
* VAST 3.0 Spec: http://www.iab.com/wp-content/uploads/2015/06/VASTv3_0.pdf
* VAST 4.0 Spec: 
  * http://www.iab.com/wp-content/uploads/2016/01/VAST_4-0_2016-01-21.pdf
  * https://www.iab.com/wp-content/uploads/2016/04/VAST4.0_Updated_April_2016.pdf
* VAST 4.1 Spec:
  * https://iabtechlab.com/wp-content/uploads/2018/11/VAST4.1-final-Nov-8-2018.pdf
* [VAST Samples](https://github.com/InteractiveAdvertisingBureau/VAST_Samples)

## Install

Install library through composer:

```
composer require sokil/php-vast
```

## Quick start

```php
// create document
$factory = new \Sokil\Vast\Factory();
$document = $factory->create('2.0');
// or, if you have at least PHP5.4
$document = (new \Sokil\Vast\Factory())->create('2.0');
// creating through Document::create and other factory methods are now deprecated:
$document = \Sokil\Vast\Document::create('2.0');

// insert Ad section
$ad1 = $document
    ->createInLineAdSection()
    ->setId('ad1')
    ->setAdSystem('Ad Server Name')
    ->setAdTitle('Ad Title')
    ->addImpression('http://ad.server.com/impression', 'imp1');

// create creative for ad section
$linearCreative = $ad1
    ->createLinearCreative()
    ->setDuration(128)
    ->setVideoClicksClickThrough('http://entertainmentserver.com/landing')
    ->addVideoClicksClickTracking('http://ad.server.com/videoclicks/clicktracking')
    ->addVideoClicksCustomClick('http://ad.server.com/videoclicks/customclick')
    ->addTrackingEvent('start', 'http://ad.server.com/trackingevent/start')
    ->addTrackingEvent('pause', 'http://ad.server.com/trackingevent/stop');
    
// add 100x100 media file
$linearCreative
    ->createMediaFile()
    ->setProgressiveDelivery()
    ->setType('video/mp4')
    ->setHeight(100)
    ->setWidth(100)
    ->setBitrate(2500)
    ->setUrl('http://server.com/media1.mp4');

// add 200x200 media file
$linearCreative
    ->createMediaFile()
    ->setProgressiveDelivery()
    ->setType('video/mp4')
    ->setHeight(200)
    ->setWidth(200)
    ->setBitrate(2500)
    ->setUrl('http://server.com/media2.mp4');
    
// get dom document
$domDocument = $document->toDomDocument();

// get XML string
echo $document;
```

This will generate:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<VAST version="2.0">
    <Ad id="ad1">
        <InLine>
            <AdSystem>Ad Server Name</AdSystem>
            <AdTitle><![CDATA[Ad Title]]></AdTitle>
            <Impression id="imp1"><![CDATA[http://ad.server.com/impression]]></Impression>
            <Creatives>
                <Creative>
                    <Linear>
                        <Duration>00:02:08</Duration>
                        <VideoClicks>
                            <ClickThrough><![CDATA[http://entertainmentserver.com/landing]]></ClickThrough>
                            <ClickTracking><![CDATA[http://ad.server.com/videoclicks/clicktracking]]></ClickTracking>
                            <CustomClick><![CDATA[http://ad.server.com/videoclicks/customclick]]></CustomClick>
                        </VideoClicks>
                        <TrackingEvents>
                            <Tracking event="start"><![CDATA[http://ad.server.com/trackingevent/start]]></Tracking>
                            <Tracking event="pause"><![CDATA[http://ad.server.com/trackingevent/stop]]></Tracking>
                        </TrackingEvents>
                        <MediaFiles>
                            <MediaFile delivery="progressive" type="video/mp4" height="100" width="100" bitrate="2500">
                                <![CDATA[http://server.com/media1.mp4]]>
                            </MediaFile>
                            <MediaFile delivery="progressive" type="video/mp4" height="200" width="200" bitrate="2500">
                                <![CDATA[http://server.com/media2.mp4]]>
                            </MediaFile>
                        </MediaFiles>
                    </Linear>
                </Creative>
            </Creatives>
        </InLine>
    </Ad>
</VAST>
```


CHANGELOG.md
## 0.5.4 (2019-08-28)
  * add custom tag
## 0.5.3 (2019-05-06)
  * UniversalAdId added due to VAST 4.0 specification. UniversalAdId required in response beginning with VAST 4.0 (paragraph 3.7.1)

## 0.5.2 (2019-03-23)
  * Add offset to progress event tracking. See specification 2.3.2.3 "Progress Event".

## 0.5.1 (2018-10-31)
  * Allow specify `adParameter` and `apiFramework`. 

## 0.4.8 (2018-06-29)
  * Fixed 'Document::getAdSections()'. Previously root DOM element was set to first child which may be text node. Now it points to `InLine` or `Wrapper` element.
  * Now may be specified id of `Impression` in `AbstractAdNode::addImpression`

## 0.4.7 (2018-06-29)
  * Fixed `AbstractNode::getValuesOfArrayNode`. Affected  `AbstractAdNode::getErrors()` and `AbstractAdNode::getImpressions()` if multiple nodes in array found. Array always contain first node instead of values of all nodes.

## 0.4.6 (2018-04-05)
  * Fix adding Extension to Ad\InLine section
  * Now Extensions can be created for Ad\Wrapper also

## 0.4.5 (2018-02-22)
  * Method `AbstractAdNode::getSequence` now return int instead of numeric string 

## 0.4.4 (2018-01-25)
  * Allow Ad sequence

## 0.4.2 (2017-11-02)
  * Returned method `AbstractAdNode::setImpression` and marked deprecated

## 0.4 (2017-07-26)
  * Move to PSR-4
  * Added `Factory` to create document. Factory methods in `Document` are deprecated
  * Added `Error` tag to `VAST`, `InLine` and `Wrapper`
  * Allowed to add multiple `Impression` to `InLine` and `Wrapper`
  * Removed method `AbstractAdNode::setImpression` 
  * Classes moved. Check your extends

## 0.3 (2016-03-24)
  * Add support of Wrapper's VASTAdTagURI element
  * Private methods and properties now without underscores


src/Creative/AbstractLinearCreative.php
<?php

namespace OpiyOrg\Vast\Creative;

use OpiyOrg\Vast\ElementBuilder;

abstract class AbstractLinearCreative extends AbstractCreative
{
    /**
     * @var ElementBuilder
     */
    protected $vastElementBuilder;

    /**
     * this event should be used to indicate when the player considers that it has loaded
     * and buffered the creative’s media and assets either fully or to the extent that it is ready to play the media.
     */
    const EVENT_TYPE_LOADED = 'loaded';

    /**
     * not to be confused with an impression, this event indicates that an individual creative
     * portion of the ad was viewed. An impression indicates the first frame of the ad was displayed; however
     * an ad may be composed of multiple creative, or creative that only play on some platforms and not
     * others. This event enables ad servers to track which ad creative are viewed, and therefore, which
     * platforms are more common.
     */
    const EVENT_TYPE_CREATIVEVIEW = 'creativeView';
                     
    /**
     * this event is used to indicate that an individual creative within the ad was loaded and playback
     * began. As with creativeView, this event is another way of tracking creative playback.
     */
    const EVENT_TYPE_START = 'start';
    
    // the creative played for at least 25% of the total duration.
    const EVENT_TYPE_FIRSTQUARTILE = 'firstQuartile';
    
    // the creative played for at least 50% of the total duration.
    const EVENT_TYPE_MIDPOINT = 'midpoint';
    
    // the creative played for at least 75% of the duration.
    const EVENT_TYPE_THIRDQUARTILE = 'thirdQuartile';
    
    // The creative was played to the end at normal speed.
    const EVENT_TYPE_COMPLETE = 'complete';
    
    // the user activated the mute control and muted the creative.
    const EVENT_TYPE_MUTE = 'mute';
    
    // the user activated the mute control and unmuted the creative.
    const EVENT_TYPE_UNMUTE = 'unmute';
    
    // the user clicked the pause control and stopped the creative.
    const EVENT_TYPE_PAUSE = 'pause';
    
    // the user activated the rewind control to access a previous point in the creative timeline.
    const EVENT_TYPE_REWIND = 'rewind';
    
    // the user activated the resume control after the creative had been stopped or paused.
    const EVENT_TYPE_RESUME = 'resume';
    
    // the user activated a control to extend the video player to the edges of the viewer’s screen.
    const EVENT_TYPE_FULLSCREEN = 'fullscreen';
    
    // the user activated the control to reduce video player size to original dimensions.
    const EVENT_TYPE_EXITFULLSCREEN = 'exitFullscreen';
    
    // the user activated a control to expand the creative.
    const EVENT_TYPE_EXPAND = 'expand';
    
    // the user activated a control to reduce the creative to its original dimensions.
    const EVENT_TYPE_COLLAPSE = 'collapse';
    
    /**
     * The user activated a control that launched an additional portion of the
     * creative. The name of this event distinguishes it from the existing “acceptInvitation” event described in
     * the 2008 IAB Digital Video In-Stream Ad Metrics Definitions, which defines the “acceptInivitation”
     * metric as applying to non-linear ads only. The “acceptInvitationLinear” event extends the metric for use
     * in Linear creative.
     */
    const EVENT_TYPE_ACCEPTINVITATIONLINEAR = 'acceptInvitationLinear';
    
    /**
     * The user clicked the close button on the creative. The name of this event distinguishes it
     * from the existing “close” event described in the 2008 IAB Digital Video In-Stream Ad Metrics
     * Definitions, which defines the “close” metric as applying to non-linear ads only. The “closeLinear” event
     * extends the “close” event for use in Linear creative.
     * Available in VAST v.3, not available in VAST v.4
     */
    const EVENT_TYPE_CLOSELINEAR = 'closeLinear';
    
    // the user activated a skip control to skip the creative, which is a
    // different control than the one used to close the creative.
    const EVENT_TYPE_SKIP = 'skip';
    
    /**
     * the creative played for a duration at normal speed that is equal to or greater than the
     * value provided in an additional attribute for offset . Offset values can be time in the format
     * HH:MM:SS or HH:MM:SS.mmm or a percentage value in the format n% . Multiple progress ev
     */
    const EVENT_TYPE_PROGRESS = 'progress';

    /**
     * Dom Element of <Creative></Creative>
     *
     * @var \DOMElement
     */
    private $linearCreativeDomElement;

    /**
     * @var \DOMElement
     */
    private $videoClicksDomElement;

    /**
     * @var \DOMElement
     */
    private $trackingEventsDomElement;

    /**
     * @param \DOMElement $linearCreativeDomElement
     * @param ElementBuilder $vastElementBuilder
     */
    public function __construct(\DOMElement $linearCreativeDomElement, ElementBuilder $vastElementBuilder)
    {
        $this->linearCreativeDomElement = $linearCreativeDomElement;
        $this->vastElementBuilder = $vastElementBuilder;
    }

    /**
     * Dom Element of <Creative></Creative>
     *
     * @return \DOMElement
     */
    protected function getDomElement()
    {
        return $this->linearCreativeDomElement;
    }

    /**
     * List of allowed events
     *
     * @return array
     */
    public static function getEventList()
    {
        return array(
            self::EVENT_TYPE_LOADED,
            self::EVENT_TYPE_CREATIVEVIEW,
            self::EVENT_TYPE_START,
            self::EVENT_TYPE_FIRSTQUARTILE,
            self::EVENT_TYPE_MIDPOINT,
            self::EVENT_TYPE_THIRDQUARTILE,
            self::EVENT_TYPE_COMPLETE,
            self::EVENT_TYPE_MUTE,
            self::EVENT_TYPE_UNMUTE,
            self::EVENT_TYPE_PAUSE,
            self::EVENT_TYPE_REWIND,
            self::EVENT_TYPE_RESUME,
            self::EVENT_TYPE_FULLSCREEN,
            self::EVENT_TYPE_EXITFULLSCREEN,
            self::EVENT_TYPE_EXPAND,
            self::EVENT_TYPE_COLLAPSE,
            self::EVENT_TYPE_ACCEPTINVITATIONLINEAR,
            self::EVENT_TYPE_CLOSELINEAR,
            self::EVENT_TYPE_SKIP,
            self::EVENT_TYPE_PROGRESS,
        );
    }

    /**
     * Get VideoClicks DomElement
     *
     * @return \DOMElement
     */
    protected function getVideoClicksDomElement()
    {
        // create container
        if (!empty($this->videoClicksDomElement)) {
            return $this->videoClicksDomElement;
        }
        
        $this->videoClicksDomElement = $this->linearCreativeDomElement->getElementsByTagName('VideoClicks')->item(0);
        if (!empty($this->videoClicksDomElement)) {
            return $this->videoClicksDomElement;
        }
        
        $this->videoClicksDomElement = $this->linearCreativeDomElement->ownerDocument->createElement('VideoClicks');
        $this->linearCreativeDomElement
            ->getElementsByTagName('Linear')
            ->item(0)
            ->appendChild($this->videoClicksDomElement);
        
        return $this->videoClicksDomElement;
    }
    
    /**
     * Add click tracking url
     *
     * @param string $url
     * @return $this
     */
    public function addVideoClicksClickTracking($url)
    {
        // create ClickTracking
        $clickTrackingDomElement = $this->getDomElement()->ownerDocument->createElement('ClickTracking');
        $this->getVideoClicksDomElement()->appendChild($clickTrackingDomElement);

        // create cdata
        $cdata = $this->getDomElement()->ownerDocument->createCDATASection($url);
        $clickTrackingDomElement->appendChild($cdata);

        return $this;
    }
    
    /**
     * Add custom click url
     *
     * @param string $url
     * @return $this
     */
    public function addVideoClicksCustomClick($url)
    {
        // create CustomClick
        $customClickDomElement = $this->getDomElement()->ownerDocument->createElement('CustomClick');
        $this->getVideoClicksDomElement()->appendChild($customClickDomElement);

        // create cdata
        $cdata = $this->getDomElement()->ownerDocument->createCDATASection($url);
        $customClickDomElement->appendChild($cdata);

        return $this;
    }

    /**
     * Set video click through url
     *
     * @param string $url
     * @return $this
     */
    public function setVideoClicksClickThrough($url)
    {
        // create cdata
        $cdata = $this->getDomElement()->ownerDocument->createCDATASection($url);

        // create ClickThrough
        $clickThroughDomElement = $this->getVideoClicksDomElement()->getElementsByTagName('ClickThrough')->item(0);
        if (!$clickThroughDomElement) {
            $clickThroughDomElement = $this->getDomElement()->ownerDocument->createElement('ClickThrough');
            $this->getVideoClicksDomElement()->appendChild($clickThroughDomElement);
        }

        // update CData
        if ($clickThroughDomElement->hasChildNodes()) {
            $clickThroughDomElement->replaceChild($cdata, $clickThroughDomElement->firstChild);
        } else { // insert CData
            $clickThroughDomElement->appendChild($cdata);
        }

        return $this;
    }

    /**
     * Get TrackingEvents DomElement
     *
     * @return \DOMElement
     */
    protected function getTrackingEventsDomElement()
    {
        // create container
        if ($this->trackingEventsDomElement) {
            return $this->trackingEventsDomElement;
        }
        
        $this->trackingEventsDomElement = $this->linearCreativeDomElement
            ->getElementsByTagName('TrackingEvents')
            ->item(0);

        if ($this->trackingEventsDomElement) {
            return $this->trackingEventsDomElement;
        }
        
        $this->trackingEventsDomElement = $this->linearCreativeDomElement
            ->ownerDocument
            ->createElement('TrackingEvents');

        $this->linearCreativeDomElement
            ->getElementsByTagName('Linear')
            ->item(0)
            ->appendChild($this->trackingEventsDomElement);
        
        return $this->trackingEventsDomElement;
    }

    /**
     * @param string $event
     * @param string $url
     *
     * @return $this
     * @throws \Exception
     */
    public function addTrackingEvent($event, $url)
    {
        if (!in_array($event, $this->getEventList())) {
            throw new \Exception(sprintf('Wrong event "%s" specified', $event));
        }
        
        // create Tracking
        $trackingDomElement = $this->linearCreativeDomElement->ownerDocument->createElement('Tracking');
        $this->getTrackingEventsDomElement()->appendChild($trackingDomElement);
        
        // add event attribute
        $trackingDomElement->setAttribute('event', $event);

        // create cdata
        $cdata = $this->linearCreativeDomElement->ownerDocument->createCDATASection($url);
        $trackingDomElement->appendChild($cdata);

        return $this;
    }

    /**
     * @param string $url
     * @param int|string $offset seconds or time in format "H:m:i" or percents in format "n%"
     *
     * @return $this
     */
    public function addProgressTrackingEvent($url, $offset)
    {
        // create Tracking
        $trackingDomElement = $this->linearCreativeDomElement->ownerDocument->createElement('Tracking');
        $this->getTrackingEventsDomElement()->appendChild($trackingDomElement);

        // add event attribute
        $trackingDomElement->setAttribute('event', self::EVENT_TYPE_PROGRESS);

        // add offset attribute
        if (is_numeric($offset)) {
            $offset = $this->secondsToString($offset);
        }
        $trackingDomElement->setAttribute('offset', $offset);

        // create cdata
        $cdata = $this->linearCreativeDomElement->ownerDocument->createCDATASection($url);
        $trackingDomElement->appendChild($cdata);

        return $this;
    }

    /**
     * Convert seconds to H:m:i
     * Hours could be more than 24
     *
     * @param mixed $seconds
     *
     * @return string
     */
    protected function secondsToString($seconds)
    {
        $seconds = (int) $seconds;

        $time = array();

        // get hours
        $hours = floor($seconds / 3600);
        $time[] = str_pad($hours, 2, '0', STR_PAD_LEFT);

        // get minutes
        $seconds = $seconds % 3600;
        $time[] = str_pad(floor($seconds / 60), 2, '0', STR_PAD_LEFT);

        // get seconds
        $time[] = str_pad($seconds % 60, 2, '0', STR_PAD_LEFT);

        return implode(':', $time);
    }

    /**
     * Get id
     *
     * @return string
     */
    public function getId()
    {
        return $this->linearCreativeDomElement->getAttribute('id');
    }

    /**
     * Set 'id' attribute
     *
     * @param string $id
     *
     * @return AbstractLinearCreative
     */
    public function setId($id)
    {
        $this->linearCreativeDomElement->setAttribute('id', $id);

        return $this;
    }


    /**
     * Set 'adid' attribute
     *
     * @param string $id
     *
     * @return AbstractLinearCreative
     */
    public function setAdId($id)
    {
        $this->linearCreativeDomElement->setAttribute('adid', $id);

        return $this;
    }

}
